Add big-endian support to TFLite FlatBuffers

kun-lu20 opened the pull request:
This PR aims to achieve a holistic/generic solution for TFLite FlatBuffers endianness issue on s390x(big-endian machines). It follows the below guidelines on BE machines:

1. Provide the endianness conversion feature for constant buffers in a TFLite model, both in C++ and Python code.
2. Convert the buffers from LE(little-endian) to BE(big-endian) format when loading a TFLite model from a file.
3. Convert the buffers from BE to LE format when writing a serialized string of TFLite model to a file.
4. Keep the buffers in BE format when the model/buffer is in memory.

This PR won't cause regression on LE machines. After applying this patch, LE format serialized TFLite model files (.tflite or .bin) could be used across platforms with different endianness format.

The following test case failures could be fixed by this code change on BE machines:
 
```bash
--- In lite module:
//tensorflow/lite:signature_runner_test
//tensorflow/lite/c:c_api_signature_runner_test
//tensorflow/lite/c:c_test
//tensorflow/lite/experimental/acceleration/mini_benchmark:validator_runner_impl_test
//tensorflow/lite/experimental/acceleration/mini_benchmark:validator_runner_test
//tensorflow/lite/experimental/acceleration/mini_benchmark:validator_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/metrics:blazeface_mlkit_v1_validation_regression_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/metrics:mobilenet_float_validation_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/metrics:mobilenet_quant_validation_test
//tensorflow/lite/experimental/acceleration/mini_benchmark/model_modifier:custom_validation_embedder_test
//tensorflow/lite/python:util_test
//tensorflow/lite/testing:zip_test_expand_dims_mlir-quant
//tensorflow/lite/testing:zip_test_fully_connected_mlir-quant
//tensorflow/lite/testing:zip_test_mean_mlir-quant
//tensorflow/lite/testing:zip_test_mirror_pad_mlir-quant
//tensorflow/lite/testing:zip_test_pad_mlir-quant
//tensorflow/lite/testing:zip_test_reduce_max_mlir-quant
//tensorflow/lite/testing:zip_test_reduce_min_mlir-quant
//tensorflow/lite/testing:zip_test_reshape_mlir-quant
//tensorflow/lite/testing:zip_test_resize_bilinear_mlir-quant
//tensorflow/lite/testing:zip_test_resize_bilinear_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_resize_nearest_neighbor_mlir-quant
//tensorflow/lite/testing:zip_test_resize_nearest_neighbor_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_slice_mlir-quant
//tensorflow/lite/testing:zip_test_softmax_mlir-quant
//tensorflow/lite/testing:zip_test_softmax_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_split_mlir-quant
//tensorflow/lite/testing:zip_test_split_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_squeeze_mlir-quant
//tensorflow/lite/testing:zip_test_strided_slice_mlir-quant
//tensorflow/lite/testing:zip_test_transpose_conv_mlir-quant
//tensorflow/lite/testing:zip_test_transpose_conv_mlir-quant_xnnpack
//tensorflow/lite/testing:zip_test_transpose_mlir-quant
//tensorflow/lite/testing:zip_test_where_mlir-quant
//tensorflow/lite/tools/optimize/calibration:calibrator_test
//tensorflow/lite/tools/optimize/debugging/python:debugger_test
//tensorflow/lite/tools/optimize/python:modify_model_interface_lib_test
//tensorflow/lite/python:lite_v2_test

--- In compiler module:
//tensorflow/compiler/mlir/lite/quantization/lite:quantize_model_test
//tensorflow/compiler/mlir/lite/quantization/lite:quantize_weights_test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:constants.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:importer_test_min_max.cc.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:importer_test_min_max.cc.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:math.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:quantization.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:reshape.mlir.test
//tensorflow/compiler/mlir/lite/tests/flatbuffer2mlir:simple.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:bucketize.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:custom_op_with_tflite_op.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:depthwise_conv2d.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:depthwise_conv2d_v2.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:disable_flex_enable_builtin.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:flex_op_with_tflite_op.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:math.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:quantization.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:reshape.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:signature_def_output_override.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:simple.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:simple_with_connected_control_nodes.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:simple_with_unconnected_control_nodes.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:tf_entry_function.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:tfl_while_op.mlir.test
//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer:while_op.mlir.test
```

This PR will also fix issue #45009 and PR #57065.

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Add BE support to TFLite FlatBuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Fix the PyLint format issue

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo,

Hope all is well.

Since this PR will fix the TFLite FlatBuffers endianness issue discussed in https://github.com/tensorflow/tensorflow/issues/45009, could you please review it when you have some time?

Thank you very much!

kun-lu20 commented:
Hi @MeghnaNatraj ,

Could you please review this PR when you have some time?

Thank you very much!

terryheo commented on tensorflow/compiler/mlir/lite/BUILD:
```diff
@@ -890,6 +890,7 @@ tf_native_cc_binary(
     srcs = ["flatbuffer_to_string.cc"],
     deps = [
         "//tensorflow/lite/schema:schema_fbs_with_reflection",
+        ":tensorflow_lite",
```
Why is this needed?

kun-lu20 replied:
This line was added here so that `FlatBufferModel::ByteSwapSerializedModel()` could be invoked in `tensorflow/compiler/mlir/lite/flatbuffer_to_string.cc`. 

The conversion could fix a series of `//tensorflow/compiler/mlir/lite/tests/mlir2flatbuffer` test cases in `compiler` module on BE machines.

kun-lu20 replied:
Yes, I'll improve the code to use `model_builder` only. Thanks!

kun-lu20 replied:
Done.

terryheo commented on tensorflow/lite/core/model_builder.cc:
```diff
@@ -90,6 +98,114 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
+  ByteSwapTFLiteModel(input_model);
+}
+
+void FlatBufferModel::ByteSwapBuffer(int8_t tensor_type, size_t buffer_size,
```
Is the target always writable?
What happens if the target memory is mmaped ?

kun-lu20 replied:
Actually `FlatBufferModel::ByteSwapFlatBufferModel()` will allocate a new `FlatBufferModel` instance for byte swapping and return its handler.

terryheo replied:
Then will it increase memory consumption?

Random thoughts, how about creating a separate tool to update LE buffers to BE buffers?

kun-lu20 replied:
Yes, it will increase memory consumption on BE platforms, but only during the model loading/saving phase, so I think it will not incur big impact to the performance.

Creating a separate tool will also work, but it will bring extra steps such as using the tool to convert the .tflite files on BE platforms before running TensorFlow code. I think embedding the byte swapping code in TensorFlow code base would be more convenient for the users, thus they can use the same steps and .tflite files across LE and BE platforms, without the need to mark which .tflite files are in which endianness format or do the conversion manually. The endianness issue becomes transparent to the users.

terryheo commented on tensorflow/lite/tools/serialization/writer_lib.cc:
```diff
@@ -69,6 +70,11 @@ TfLiteStatus WriteImpl(const std::string& filename, void* data, size_t size) {
   FILE* fp = fopen(filename.c_str(), "wb");
   if (!fp) return kTfLiteError;
 
+  if (!FLATBUFFERS_LITTLEENDIAN){
```
Who does define this?

kun-lu20 replied:
`FLATBUFFERS_LITTLEENDIAN` is defined in the `flatbuffers` header file here: https://github.com/google/flatbuffers/blob/6f895f54c25aa19f5d84ac6cf7fa8bc955a14e1d/include/flatbuffers/base.h#L119-L139

terryheo replied:
Isn't it better to use "#ifdef FLATBUFFERS_LITTLEENDIAN" ?

kun-lu20 replied:
Yes, I'll adjust the code accordingly. Thanks!

kun-lu20 replied:
Done.

terryheo commented on tensorflow/lite/tools/test_utils.py:
```diff
@@ -117,7 +117,7 @@ def build_mock_flatbuffer_model():
   schema_fb.TensorAddName(builder, string1_offset)
   schema_fb.TensorAddShape(builder, shape1_offset)
-  schema_fb.TensorAddType(builder, 0)
+  schema_fb.TensorAddType(builder, 3) #UINT8
```
for what?

terryheo replied:
Got it. Then `schema_fb.TensorAddType(builder, TensorType_UINT8)` ?

kun-lu20 replied:
Yes, using `TensorType_UINT8` is better. I'll update the code. Thanks!

kun-lu20 commented:
Hi @terryheo ,

Thanks for your feedback! I've replied to your comments, please take a look when you have some time.

kun-lu20 committed:
Fix the endianness issue in mini_benchmark test group on BE machines

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo ,

Could you please review this PR again? Thank you very much!

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo ,

I've solved the conflicts from master branch. Could you please do the review again when you have some time? 

Thank you!

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

terryheo commented on tensorflow/lite/core/model_builder.cc:
```diff
@@ -90,6 +98,114 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
                                       error_reporter);
 }
 
+void FlatBufferModel::ByteSwapSerializedModel(std::string* serialized_model){
```
I also want to add conditional building (`#ifdef FLATBUFFERS_LITTLEENDIAN`) for all swapping functions which are only needed for BE targets.

kun-lu20 replied:
Yes, that makes sense. I'll add the macro conditional statement for them. Thank you very much for all the valuable comments!

kun-lu20 replied:
Done.

kun-lu20 commented on tensorflow/lite/core/model_builder.cc:
```diff
@@ -90,6 +98,114 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
+    const tflite::SubGraph* subgraph = tfl_model->subgraphs()->Get(subgraph_idx);
+    for (size_t ts_idx = 0; ts_idx < subgraph->tensors()->size(); ts_idx++) {
+      const tflite::Tensor* tensor = subgraph->tensors()->Get(ts_idx);
+      if(tensor->buffer()>0 && tensor->buffer()<tfl_model->buffers()->size()){
```
I see. Will adjust the code to do swap only once. Thanks!

kun-lu20 replied:
Both the C++ code and Python code were updated to ensure the swapping would be done only once.

kun-lu20 committed:
Update the code change as per the review comments:
1. Do byte swapping only once on buffers
2. Add conditional building for all swapping functions
3. Use schema_fb.TensorType.UINT8 instead of value `3`

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo ,

Thanks again for your valuable comments! I've updated the PR as per your suggestions, please review it again when you have some time.

kun-lu20 committed:
Use schema_fb.TensorType in byte_swap Python code

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

gbaned commented:
Hi @terryheo Can you please review this PR ? Thank you!

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo ,

Happy New Year!

I've solved the conflicts from the master branch. Could you please review this PR again? Thank you very much!

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo ,

I've solved the conflicts from the master branch. Could you please take a look at this PR?

Thank you very much!

terryheo commented on tensorflow/compiler/mlir/lite/flatbuffer_to_string.cc:
```diff
@@ -137,6 +140,9 @@ int main(int argc, char** argv) {
 
   std::string serialized_model;
   if (tflite::ReadAndVerify(argv[1], &serialized_model)) return 1;
+  #if FLATBUFFERS_LITTLEENDIAN == 0
```
We don't indent for "#if" "#endif". Plz update all.

kun-lu20 replied:
Done. Thanks!

terryheo commented on tensorflow/lite/core/model_builder.cc:
```diff
@@ -90,6 +100,124 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
+  }
+}
+
+void FlatBufferModel::ByteSwapTFLiteModel(const tflite::Model* tfl_model){
```
Could you run formatter? https://www.tensorflow.org/community/contribute/code_style#c_coding_style

Need a space before '{'

kun-lu20 replied:
I've run the formatter and updated the code. Thanks!

kun-lu20 committed:
Run formatter to make the code conform to Google style

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo ,

Thanks for your feedback! I've run the formatter and updated the related code. 

Please take a look when you have some time. Thanks again!

kun-lu20 commented:
Hi @terryheo ,

Hope all is well.

Could you please review this PR again? Thank you very much!

kun-lu20 commented:
Hi @terryheo ,

Could you please have a look at the code changes when you have some time?

Thank you!

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

terryheo commented on tensorflow/lite/core/model_builder.cc:
```diff
@@ -94,6 +105,124 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
+    case tflite::TensorType_FLOAT32:
+    case tflite::TensorType_INT32:
+    case tflite::TensorType_UINT32:
+    case tflite::TensorType_COMPLEX64: {
```
TensorType_COMPLEX64 is 64bit

terryheo replied:
More preciously it consists with two 32bit values.

kun-lu20 replied:
Thanks @terryheo !

If I understood correctly, a TensorType_COMPLEX64 data is 64-bit but it's composed of two 32-bit integer numbers, so it's handled along with other 32-bit data types.

Please correct me if I was wrong. Thanks again!

terryheo replied:
Oh, I see. You're right.

terryheo commented on tensorflow/lite/core/model_builder.cc:
```diff
@@ -94,6 +105,124 @@ std::unique_ptr<FlatBufferModel> FlatBufferModel::VerifyAndBuildFromBuffer(
+    case tflite::TensorType_INT64:
+    case tflite::TensorType_FLOAT64:
+    case tflite::TensorType_UINT64:
+    case tflite::TensorType_COMPLEX128: {
```
TensorType_COMPLEX128 is 128bit

kun-lu20 replied:
The reason is the same as the above TensorType_COMPLEX64.

terryheo reviewed (commented):
Mostly LGTM

terryheo reviewed (approved):
LGTM.

If there is  an internal test error with this PR, I'll ask you to fix it.

kun-lu20 commented:
Sure, will do. Thanks @terryheo !

terryheo commented on tensorflow/lite/core/model_builder.h:
```diff
@@ -155,6 +155,33 @@ class FlatBufferModel {
+  /// Byte swap the buffers field of a TFLite ModelT instance in place.
+  static void ByteSwapTFLiteModelT(tflite::ModelT* tfl_modelt);
+
+  /// Convert the TFLite buffers field between LE and BE format in a 
```
This file has some unnecessary ending white spaces. Could you remove all?

kun-lu20 replied:
Sure

kun-lu20 replied:
Hi @terryheo ,

I've used `clang-format` to remove all unnecessary ending white spaces from this file. Also I've fixed another format issue in `tensorflow/lite/tools/serialization/BUILD` according to the `Code Check - Changed Files` results.

Could you please take a look?  Thank you very much!

kun-lu20 committed:
Remove unnecessary ending white spaces in model_builder.h

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

terryheo reviewed (approved):
LGTM

kun-lu20 commented:
Thanks @terryheo

kun-lu20 committed:
Merge branch 'master' into add_big_endian_support_to_tflite_flatbuffers

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented:
Hi @terryheo ,

I've fixed the conflicts from the master branch. Please take a look when you have some time. 

BTW, I found that the latest result from `feedback/copybara` showed the internal checks failed on `2023-02-16`. Are there any code/BUILD files I need to change? 

Thanks!

kun-lu20 commented:
Hi @terryheo ,

I saw all the checks have passed. Wondering if it's possible for this PR to be merged?

Thank you very much!

terryheo commented on tensorflow/lite/tools/flatbuffer_utils.py:
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
 
 
+def byte_swap_buffer_content(buffer, chunksize, from_endiness, to_endiness):
+  """"Helper function for byte-swapping the buffers field.
```
"""Helper function for byte-swapping the buffers field."""

1. make it one line
2. there are 4 double quotes.

kun-lu20 replied:
Updated. Thanks!

terryheo commented on tensorflow/lite/tools/flatbuffer_utils.py:
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
+
+  Args:
+    model: TFLite model object of from_endiness format.
+  
```
you need to describe every arguments

kun-lu20 replied:
Added description for the remaining arguments. Thanks!

terryheo commented on tensorflow/lite/tools/flatbuffer_utils.py:
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
+
+
+def byte_swap_tflite_model_obj(model, from_endiness, to_endiness):
+  """"Byte swaps the buffers field in a TFLite model.
```
plz use 3 double quotes instead of 4

kun-lu20 replied:
Updated. Thanks!

terryheo commented on tensorflow/lite/tools/flatbuffer_utils.py:
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
+
+
+def byte_swap_tflite_buffer(tflite_model, from_endiness, to_endiness):
+  """"Generates a new model byte array after byte swapping its buffers field.
```
use 3 double quotes

kun-lu20 replied:
Updated. Thanks!

kun-lu20 committed:
Update the function descriptions in flatbuffer_utils.py

Signed-off-by: Kun-Lu <kun.lu@ibm.com>

kun-lu20 commented on tensorflow/lite/tools/flatbuffer_utils.py:
```diff
@@ -274,6 +281,78 @@ def xxd_output_to_object(input_cc_file):
+  """"Generates a new model byte array after byte swapping its buffers field.
+
+  Args:
+    tflite_model: TFLite flatbuffer in a byte array of from_endiness format.
```
Description added. Thanks!

kun-lu20 commented:
Hi @terryheo ,

Thanks for pointing out the above issues. I've addressed them and please take a look when you have a chance. Thanks again!

terryheo reviewed (approved):
LGTM
